lib/gears/matrix.lua

```diff
@@ -150,6 +176,43 @@ function matrix:equals(other)
     return true
 end
 
+--- Merge a complete or partial matrix into the current one
+-- @param other A matrix of element table (const)
+function matrix:merge(other)
+    for _, v in ipairs { "xx", "xy", "x0", "yx", "yy", "y0" } do
+        self[v] = other[v] and other[v] or self[v]
+    end
+end
+
+--- Save the matrix state into a new memento
+function matrix:save()
+    if not rawget(self, "memento") then
+        rawset(self, "memento", {})
+    end
+
+    table.insert(self.memento, {
+        xx = self.xx, xy = self.xy, x0 = self.x0,
+        yx = self.yx, yy = self.yy, y0 = self.y0
+    })
+end
+
+--- Restore the current state of the matrix to the previous memento
+function matrix:restore()
+    if rawget(self, "memento") then
+        local pop = self.memento[#self.memento]
+        if pop then
+            table.remove(self.memento, #self.memento)
+            self:merge(pop)
+        end
+    end
+end
+
+--- When adding transformations to the matrix, retroactively apply them
+-- @tparam boolean mutate Enable mutation
+function matrix:set_mutating(mutate)
+    rawset(self, "_automerge", mutate)
+end
+
```

Comment:
I don't like all of this change to `gears.matrix`. I made matrices immutable on purpose so that e.g. I don't have to return a copy from functions in `wibox.hierarchy`, but can just return the internal matrix. If a matrix is immutable, you don't need to save/restore it, but can just keep the old instance around (in a stack?).

Let's see what you are using this for in the other commits...